src/server/service/kv.rs

```diff
@@ -904,48 +890,58 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
 }
 
 fn response_batch_commands_request<F>(
-    executor: &tokio_threadpool::Sender,
     id: u64,
     resp: F,
     tx: Sender<(u64, BatchCommandsResponse_Response)>,
     timer: HistogramTimer,
-    thread_load: Arc<ThreadLoad>,
 ) where
     F: Future<Item = BatchCommandsResponse_Response, Error = ()> + Send + 'static,
 {
-    let executor1 = executor.clone();
     let f = resp.and_then(move |resp| {
-        if tx.send((id, resp)).is_err() {
+        if tx.send_and_notify((id, resp)).is_err() {
             error!("KvService response batch commands fail");
             return Err(());
         }
         timer.observe_duration();
-        if let Some(notifier) = tx.get_notifier() {
-            if thread_load.in_heavy_load() {
-                let _ = executor1.spawn(
-                    GLOBAL_TIMER_HANDLE
-                        .delay(Instant::now() + DELAY_DURATION)
-                        .map_err(|e| error!("batch commands delay error"; "err" => ?e))
-                        .inspect(move |_| notifier.notify()),
-                );
-            } else {
-                notifier.notify();
-            }
-        }
         Ok(())
     });
-    let _ = executor.spawn(f);
+    poll_future_notify(f);
+}
+
+// BatchCommandsNotify is used to make business pool notifiy completion queues directly.
+struct BatchCommandsNotify<F>(Arc<Mutex<Option<Spawn<F>>>>);
+impl<F> Clone for BatchCommandsNotify<F> {
+    fn clone(&self) -> BatchCommandsNotify<F> {
+        BatchCommandsNotify(Arc::clone(&self.0))
+    }
+}
+impl<F> Notify for BatchCommandsNotify<F>
+where
+    F: Future<Item = (), Error = ()> + Send + 'static,
+{
+    fn notify(&self, id: usize) {
+        let n = Arc::new(self.clone());
+        let mut s = self.0.lock().unwrap();
+        match s.as_mut().map(|spawn| spawn.poll_future_notify(&n, id)) {
```

Comment:
Will `notify ` be called in the `poll_future_notify ` recursively?

Reply:
`poll_future_notify` will call `notify` once, no recursion.